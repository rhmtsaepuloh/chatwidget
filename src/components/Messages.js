/* eslint-disable react/jsx-no-target-blank */
/* eslint-disable jsx-a11y/iframe-has-title */
/* eslint-disable no-unused-expressions */
/* eslint-disable jsx-a11y/img-redundant-alt */
import React from "react";
import isEmpty from "lodash/isEmpty";
import Header from "./Header";
import PreviewImage from "./PreviewImage";
import Document from "./Document";
import DocumentProgress from "./DocumentProgress";
import TextMessagesProgress from "./TextMessagesProgress";
import TextMessages from "./TextMessages";
import ButtonMessages from "./ButtonMessages";
import ButtonMessagesProgress from "./ButtonMessagesProgress";
import Carousel from "./Carousel";

const Messages = ({
  urlImagePreview,
  onGoBack,
  displayEmoji,
  messagesList,
  uniqueUserId,
  messagesEndRef,
  onDeleteImage,
  logo,
  file,
  hexColor,
  progressUpload,
  cancelUpload,
  onPostBack,
  downloadFile,
  recomendedData,
}) => {
  const renderMessages = (item, hexColor, cancelUpload, progressUpload) => {
    if (item.type === "buttons") {
      return (
        <>
          {item.messageUniqueId === "progress" ? (
            <ButtonMessagesProgress
              message={item.payload.text}
              timestamp={item.timestamp}
              item={item}
              onPostBack={onPostBack}
            />
          ) : (
            <ButtonMessages
              message={item.payload.text}
              timestamp={item.timestamp}
              item={item}
              onPostBack={onPostBack}
            />
          )}
        </>
      );
    } else if (item.type === "file_attachment") {
      const fileName = item.payload.file_name
      return (
        <>
          {item.messageUniqueId === "progress" ? (
            <DocumentProgress
              extension={item.payload.ext}
              color={hexColor}
              progressUpload={progressUpload}
              sourceImage={item.payload.url}
              timestamp={item.timestamp}
              text={fileName && fileName.length > 35 ? (fileName.split(".").shift()).substring(0,35) + "....." + fileName.split(".").pop() : fileName}
              caption={item.payload.caption}
              cancelUpload={cancelUpload}
            />
          ) : (
            <Document
              extension={item.payload.url}
              color={hexColor}
              timestamp={item.timestamp}
              text={fileName && fileName.length > 35 ? (fileName.split(".").shift()).substring(0,35) + "....." + fileName.split(".").pop() : fileName}
              size={item.payload.size}
              caption={item.payload.caption}
              downloadFile={downloadFile}
            />
          )}
        </>
      );
    } else {
      return (
        <>
          {item.messageUniqueId === "progress" ? (
            <TextMessagesProgress
              message={item.message}
              timestamp={item.timestamp}
            />
          ) : (
            <TextMessages message={item.message} timestamp={item.timestamp} />
          )}
        </>
      );
    }
  };

  return isEmpty(urlImagePreview) ? (
    <>
      <Header onGoBack={onGoBack} />
      <div
        className='wrapper-message'
        style={{
          height: displayEmoji ? "60%" : "auto",
          zIndex: 1,
          marginBottom: isEmpty(recomendedData) ? 75 : 125,
        }}>
        {messagesList.map((message, key) => {
          return (
            <React.Fragment key={key}>
              {isEmpty(message.data) ? null : (
                <div className='date bg-gray-100 mx-auto w-32 clear-both rounded-sm'>
                  <p className='text-center'>{message.date}</p>
                </div>
              )}
              {message.data.map((item, i) => {
                return (
                  <React.Fragment key={i}>
                    {item.user.unique_user_id === uniqueUserId ? (
                      <div className='me bubble-messages' key={i}>
                        {renderMessages(
                          item,
                          hexColor,
                          cancelUpload,
                          progressUpload
                        )}
                      </div>
                    ) : item.type === "carousel" ? (
                      <React.Fragment key={i}>
                        <Carousel
                          onPostBack={onPostBack}
                          logoAdmin={logo}
                          item={item}
                        />
                      </React.Fragment>
                    ) : (
                      <div className='him flex' key={i}>
                        <div className='admin bubble-messages bg-white m-0 p-0'>
                          {renderMessages(
                            item,
                            hexColor,
                            cancelUpload,
                            progressUpload
                          )}
                        </div>
                      </div>
                    )}
                  </React.Fragment>
                );
              })}
              <div className='bottom-messages bubble-messages' />
              <br />
            </React.Fragment>
          );
        })}
        <div ref={messagesEndRef} />
      </div>
    </>
  ) : (
    <>
      <PreviewImage
        src={urlImagePreview}
        onDeleteImage={onDeleteImage}
        file={file}
        color={hexColor}
      />
      <div ref={messagesEndRef} />
    </>
  );
};

export default Messages;
