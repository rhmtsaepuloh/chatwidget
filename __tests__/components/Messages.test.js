/* eslint-disable react/react-in-jsx-scope */
import React from "react";
import { shallow } from "enzyme";
import Messages from "../../src/components/Messages";

const props = {
  onGoBack: jest.fn(),
  messagesList: [
    {
      date: "Hari ini",
      data: [
        {
          id: 420871933,
          message: "cek",
          messageUniqueId: "bzZvNGNu7ezlzTBdPSlE1602126125",
          payload: {},
          timestamp: "10:02",
          type: "text",
          user: {
            avatar_url:
              "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
            fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
            phone_number: null,
            unique_user_id: "user_36d7cf432e381bbf0fb17e7a4",
          },
        },
        {
          id: 42087193,
          message:
            "[file] http://chataja-widget-be-dev.vsan-apps.playcourt.id/public/1602128346174-js.png [/file]",
          messageUniqueId: "WPK4cjF2hdfxVlwVuUD31602128347",
          payload: {
            caption: "tambahkan",
            encryption_key: "",
            file_name: "js.png",
            pages: 1,
            size: 4689,
            url:
              "http://chataja-widget-be-dev.vsan-apps.playcourt.id/public/1602128346174-js.png",
          },
          timestamp: "10:39",
          type: "file_attachment",
          user: {
            avatar_url:
              "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
            fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
            phone_number: null,
            unique_user_id: "user_36d7cf432e381bbf0fb17e7a4f",
          },
        },
      ],
    },
  ],
  messagesEndRef: null,
  onDeleteImage: jest.fn(),
  logo:
    "https://d1edrlpyc25xu0.cloudfront.net/image/upload/t5XWp2PBRt/1510641299-default_user_avatar.png",
  file: {},
  hexColor: "#000",
  progressUpload: 0,
  uniqueUserId: "user_36d7cf432e381bbf0fb17e7a4",
};

describe("render correctly messages components", () => {
  test("renders if displayEmoji is true and urlImagePreview is available", () => {
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={true}
        urlImagePreview={
          "https://chatwidget-chataja.s3.ap-southeast-1.amazonaws.com/0fcedfb7-1861-42c0-9515-13603ce62901.png"
        }
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders if displayEmoji is false and urlImagePreview is not available", () => {
    const wrapper = shallow(
      <Messages {...props} displayEmoji={false} urlImagePreview={null} />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders if messages data is null", () => {
    const messagesList = [
      {
        date: "Hari ini",
        data: [],
      },
    ];
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        messagesList={messagesList}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders messages on user", () => {
    const wrapper = shallow(
      <Messages {...props} displayEmoji={false} urlImagePreview={null} />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders messages on admin", () => {
    const uniqueUserId = null;
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        uniqueUserId={uniqueUserId}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders if file upload is progress on user", () => {
    const messagesList = [
      {
        date: "Hari ini",
        data: [
          {
            id: 420871933,
            message: "cek",
            messageUniqueId: "progress",
            payload: {},
            timestamp: "10:02",
            type: "file_attachment",
            user: {
              avatar_url:
                "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
              fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
              phone_number: null,
              unique_user_id: "user_36d7cf432e381bbf0fb17e7a4",
            },
          },
        ],
      },
    ];
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        messagesList={messagesList}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders if file upload is progress on admin", () => {
    const uniqueUserId = null;
    const messagesList = [
      {
        date: "Hari ini",
        data: [
          {
            id: 420871933,
            message: "cek",
            messageUniqueId: "progress",
            payload: {},
            timestamp: "10:02",
            type: "file_attachment",
            user: {
              avatar_url:
                "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
              fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
              phone_number: null,
              unique_user_id: "user_36d7cf432e381bbf0fb17e7a4",
            },
          },
        ],
      },
    ];
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        messagesList={messagesList}
        uniqueUserId={uniqueUserId}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders messages is progress on user", () => {
    const messagesList = [
      {
        date: "Hari ini",
        data: [
          {
            id: 420871933,
            message: "cek",
            messageUniqueId: "progress",
            payload: {},
            timestamp: "10:02",
            type: "text",
            user: {
              avatar_url:
                "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
              fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
              phone_number: null,
              unique_user_id: "user_36d7cf432e381bbf0fb17e7a4",
            },
          },
        ],
      },
    ];
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        messagesList={messagesList}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  test("renders messages is progress on admin", () => {
    const uniqueUserId = null;
    const messagesList = [
      {
        date: "Hari ini",
        data: [
          {
            id: 420871933,
            message: "cek",
            messageUniqueId: "progress",
            payload: {},
            timestamp: "10:02",
            type: "text",
            user: {
              avatar_url:
                "https://file-manager.chataja.co.id/rails/active_storage/representations/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBbHVSIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--80de9c37cf322d78d2c909ff8f2d4245a2d3256d/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaDdCam9RWVhWMGIxOXZjbWxsYm5SVSIsImV4cCI6bnVsbCwicHVyIjoidmFyaWF0aW9uIn19--9fbd58a8279dfc2b55a934aacd1a224d5bfb00a8/1593483285.png",
              fullname: "user_36d7cf432e381bbf0fb17e7a4fab57af",
              phone_number: null,
              unique_user_id: "user_36d7cf432e381bbf0fb17e7a4",
            },
          },
        ],
      },
    ];
    const wrapper = shallow(
      <Messages
        {...props}
        displayEmoji={false}
        urlImagePreview={null}
        messagesList={messagesList}
        uniqueUserId={uniqueUserId}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
