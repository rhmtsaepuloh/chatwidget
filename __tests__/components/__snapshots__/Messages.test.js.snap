// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly messages components renders if displayEmoji is false and urlImagePreview is not available 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders if displayEmoji is true and urlImagePreview is available 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders if file upload is progress on admin 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders if file upload is progress on user 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders if messages data is null 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders messages is progress on admin 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders messages is progress on user 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders messages on admin 1`] = `ShallowWrapper {}`;

exports[`render correctly messages components renders messages on user 1`] = `ShallowWrapper {}`;
